server/sonar-server/src/main/java/org/sonar/server/issue/ws/TagsAction.java

```diff
@@ -58,21 +76,45 @@ public void define(WebService.NewController controller) {
       .setDescription("The size of the list to return")
       .setExampleValue("25")
       .setDefaultValue("10");
+    action.createParam(PARAM_ORGANIZATION)
+      .setDescription("Organization key")
+      .setRequired(false)
+      .setInternal(true)
+      .setExampleValue("my-org")
+      .setSince("6.4");
   }
 
   @Override
   public void handle(Request request, Response response) throws Exception {
     String query = request.param(Param.TEXT_QUERY);
     int pageSize = request.mandatoryParamAsInt("ps");
-    List<String> tags = listTags(query, pageSize);
-    writeTags(response, tags);
+    OrganizationDto organization = getOrganization(request.param(PARAM_ORGANIZATION));
+    List<String> tags = listTags(organization, query, pageSize == 0 ? Integer.MAX_VALUE : pageSize);
+    writeResponse(response, tags);
   }
 
-  private List<String> listTags(@Nullable String textQuery, int pageSize) {
-    return issueIndex.listTags(textQuery, pageSize);
+  private List<String> listTags(OrganizationDto organization, @Nullable String textQuery, int pageSize) {
+    Collection<String> issueTags = issueIndex.listTags(organization, textQuery, pageSize);
+    Collection<String> ruleTags = ruleIndex.listTags(organization, textQuery, pageSize);
+
+    SortedSet<String> result = Sets.newTreeSet();
+    result.addAll(issueTags);
+    result.addAll(ruleTags);
+    List<String> resultAsList = Lists.newArrayList(result);
```

Comment:
In order to have this algorithm working, `ruleIndex.listTags()` must return a list ordered alphabetically. I already did the comment in RuleIndex.